Google IOT Mqtt TLS sample

agross-oss opened the pull request:
This set of patches implements a full mqtt tls jwt stack for communication/authenticating with the Google IOT Cloud servers.

d3zd3z committed:
jwt: Add JSON web token library

This patch adds a JSON web token library that adds the capability
to sign JSON tokens.  This was located in subsys due to the dependency
on MBEDTLS, which resides in /ext.

Signed-off-by: David Brown <david.brown@linaro.org>
Signed-off-by: Andy Gross <andy.gross@linaro.org>

d3zd3z commented on drivers/jwt/Kconfig:
```diff
@@ -0,0 +1,31 @@
+# This is a hack.
```
Remove this, it is no longer true.

agross-linaro replied:
fixed

d3zd3z commented on ext/lib/crypto/mbedtls/configs/config-full-mqtt.h:
```diff
@@ -0,0 +1,88 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Minimal configuration for TLS 1.2 (RFC 5246) for Zephyr, implementing only
+ * a few of the most popular ciphersuites.
```
I wonder if we should be more explicit that this is about adding the support needed to communicate using MQTT. It's obvious from the filename, but not the source.

agross-linaro replied:
done

d3zd3z commented on samples/net/full_mqtt_tls/README.rst:
```diff
@@ -0,0 +1,41 @@
+
+Users will also be required to configure the following Kconfig options
+based on their Google Cloud IOT project:
+- CLOUD_CLIENT_ID - created from <telemetry>/<region>/registries/<registry name>/devices/<device name>
```
It is unclear from this comment what is wanted here. Is CLOUD_CLIENT_ID supposed to be the whole client ID string, or is it included as part of it. Same for audience, subscription and publish.

agross-linaro replied:
im going to point to the google doc that specifies what these are

agross-linaro replied:
Added google doc link.

d3zd3z commented on samples/net/full_mqtt_tls/README.rst:
```diff
@@ -0,0 +1,41 @@
+based on their Google Cloud IOT project:
+- CLOUD_CLIENT_ID - created from <telemetry>/<region>/registries/<registry name>/devices/<device name>
+- CLOUD_AUDIENCE - created from telementry without projects on front
+- CLOUD_SUBSCRIPTION - created from /devices/<device name>/config
```
This should have CONFIG in the name. There could be other things a client would want to subscribe to, and the config data is handled specially by the cloud service.

agross-linaro replied:
done

d3zd3z commented on samples/net/full_mqtt_tls/src/dhcp.c:
```diff
@@ -0,0 +1,87 @@
+/* DHCPv4 client startup. */
```
Is this file used? I thought dhcp was being handled by the net app layer.

agross-linaro replied:
no, it still has to be kicked off here.  from what i can tell at least.

d3zd3z commented on samples/net/full_mqtt_tls/src/globalsign.inc:
```diff
@@ -0,0 +1,99 @@
+/*
```
I wonder if we should document where this came from. This will have to be replaced at some point (it expires in 2021), and Google is free to change who issues their certificate. Their docs currently give a list of certificates (about 60K), which we're not really set up to handle. Having this in a sample is fine, but there will need to be a better way of handling this.

It is also important to emphasize that having this cert is an _absolute requirement_. Using TLS without verifying the certificate is pointless, and arguably worse than not bothering with TLS in the first place. Man-in-the-middle attacks become trivial without it, and there are no security assurances at all about the communication.

agross-linaro replied:
added some text to the file to indicate the source of the file.  Ideally we'd have a script that populates this, but we can look to that for the future.

d3zd3z commented on samples/net/full_mqtt_tls/src/private_info/create_keys.py:
```diff
@@ -0,0 +1,91 @@
+        subprocess.call(["openssl", "genrsa", "-out", pem_file, "2048"])
+
+    print("Generating certificate")
+    subprocess.call(["openssl", "req", "-new", "-x509", "-key", pem_file, "-out",
```
Should use subprocess.check_call so that errors aren't ignored. Likewise other uses in this file.

agross-linaro replied:
fixed all instances.  used try except blocks

d3zd3z commented on samples/net/full_mqtt_tls/src/private_info/create_keys.py:
```diff
@@ -0,0 +1,91 @@
+            "-offset", "5", "-length", "34"], stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT)
+
+        tmp = o.communicate()[0].decode('ascii').split(":")[3].lower().strip()
```
Should check o.returncode to make sure that the command ran without error.

agross-linaro replied:
fixed

d3zd3z commented on tests/lib/jwt/src/main.c:
```diff
@@ -0,0 +1,302 @@
+	JSON_OBJ_DESCR_PRIM(struct jwt_payload, iat, JSON_TOK_NUMBER),
+};
+
+static void test_json_encode(void)
```
This whole section function, and most of what is above it that it calls should just be removed. It migrated into the JWT code and will be tested as part of the JWT test. There isn't really much point in testing what is now an independent implementation.

agross-linaro replied:
done

ulfalizer commented on samples/net/full_mqtt_tls/Kconfig:
```diff
@@ -0,0 +1,35 @@
+	help
+	  Set publish topics string that matches Google Cloud project
+
+source "$(ZEPHYR_BASE)/Kconfig.zephyr"
```
Can make this just

    source "Kconfig.zephyr"

See https://github.com/zephyrproject-rtos/zephyr/pull/10487.

agross-linaro replied:
done

ulfalizer commented on drivers/jwt/Kconfig:
```diff
@@ -0,0 +1,31 @@
+	select JSON_LIBRARY
+	default n
+	help
+	  Enable creation of JWT tokens
```
Could make this

```
menuconfig JWT
	bool "Enable JWT token generation"
	select JSON_LIBRARY
	help
	  Enable creation of JWT tokens
```

`bool` symbols implicitly default to `n`, and `bool "Foo"` is a shorthand for`bool` + `prompt "Foo"`.

ulfalizer reviewed (approved):
Kconfig nits fixed.

rlubos commented:
Hello! It's really good to the sample codebase is expanding, especially with such a nice features as cloud connectivity.

I have a few generic questions though, without diving deep into details:
* I can see that the sample brings it's own implementation of MQTT (which seems very similar to the current Zephyr MQTT implementation). Is there any pariticular reason to do that? Is there anything specific with the implementation proposed?
* Why do we use here `zsock_` socket API instead of POSIX names? This will prevent this sample from any platform supporting socket offloading.
* This one is more for my own understandig. Quite recently, a convenience API to use TLS with sockets was introduced. Why don't you use it here? I know the reason might be that there's no docs yet (I know it's my fault!). Or is there any other reason the API does not fit the needs of this sample?

And finally, FYI, there's a MQTT rework happening in Zephyr on a PR #10125. It introduces new, socket based implementation of MQTT, which utilizes the TLS API as well, (though the "old" one will still exist for some time, so no worries). The sample there can connect to a TLS authenticated server, so you could have a look at it (it was tested locally, but I see no reason why shouldn't it work with a cloud server).

Thanks!

jukkar commented on drivers/jwt/Kconfig:
```diff
@@ -0,0 +1,27 @@
+menuconfig JWT
```
Why this is in driver subdirectory, could we make it a library as you are calling it a library in commit msg?

agross-linaro replied:
i actually did mean this to be moved to library.  I'll fix that.

jukkar commented on ext/lib/crypto/mbedtls/configs/config-full-mqtt.h:
```diff
@@ -0,0 +1,90 @@
+/*
```
Could you separate the JWT part from this sample, meaning that submitting the JWT in separate PR?

jukkar replied:
I would rather see the new MQTT at #10125 used as a base and this sample building on top of that PR. We have already a MQTT library that is being replaced by #10125 so I would not like to introduce yet another MQTT library even thou this is a sample.

agross-linaro replied:
I am not a fan of separating work that is required by the work I actually want to do.  github really doesn't handle this well.  If we could get immediate turnaround on JWT, that would be a compelling argument.  But in most cases this draws out things considerably.

agross-linaro replied:
and I'd be happy to pivot to any merged change.  Key word, merged.

agross-linaro replied:
I moved the JWT to subsys.  This was required due to the MBEDTLS dependency.  It has to be added after /ext directory

d3zd3z commented:
> I can see that the sample brings it's own implementation of MQTT (which seems very similar to the current Zephyr MQTT implementation). Is there any pariticular reason to do that? Is there anything specific with the implementation proposed?

Several reasons, primarily is that the MQTT change is still an open PR. This application requires TLS, and the existing mqtt code is heavily dependent on the underlying layer.

Github is quite poor at handling changes that depend on other changes, especially multiple other changes (this would need both MQTT and the TLS under socket). Changing this code to use these libraries is a good idea, but I'm not sure that should be done until both of those changes merge.

I'd much rather get the code in place, and then change it later, once these other changes get merged. Otherwise, we end up with lots of code that is unmerged, and is much more likely to rot because of other changes. Especially since the code is entirely in a sample. Our intent is definitely to keep this up to date, and try to make this code smaller and smaller over time. But, if we always have some other change that needs to go in first, the sample will never make it into the tree.

d3zd3z commented:
> I can see that the sample brings it's own implementation of MQTT (which seems very similar to the current Zephyr MQTT implementation). Is there any pariticular reason to do that? Is there anything specific with the implementation proposed?

Several reasons, primarily is that the MQTT change is still an open PR. This application requires TLS, and the existing mqtt code is heavily dependent on the underlying layer.

Github is quite poor at handling changes that depend on other changes, especially multiple other changes (this would need both MQTT and the TLS under socket). Changing this code to use these libraries is a good idea, but I'm not sure that should be done until both of those changes merge.

I'd much rather get the code in place, and then change it later, once these other changes get merged. Otherwise, we end up with lots of code that is unmerged, and is much more likely to rot because of other changes. Especially since the code is entirely in a sample. Our intent is definitely to keep this up to date, and try to make this code smaller and smaller over time. But, if we always have some other change that needs to go in first, the sample will never make it into the tree.

d3zd3z commented:
> Could you separate the JWT part from this sample, meaning that submitting the JWT in separate PR?

What benefit is there to doing this? Nothing yet depends on JWT, and github is pretty terrible at dealing with dependent pull requests.

d3zd3z commented:
> The sample there can connect to a TLS authenticated server, so you could have a look at it (it was tested locally, but I see no reason why shouldn't it work with a cloud server).

One of Google Core IoT's "features" is that they don't use TLS to authenticate the clients. That's where the JWT code comes from.

Again, moving to the new MQTT and socket/TLS api is a great idea, but lets get the order right. This is an isolated sample app that works with the code that is in tree. Let's make it work with the pending code once that pending code is merged.

jukkar commented:
> What benefit is there to doing this? Nothing yet depends on JWT, and github is pretty terrible at dealing with dependent pull requests.

Just pondering that it might get merged faster that way. The sample requires more work before it can be accepted. But we can certainly keep the JWT patch here if you wish.

dbkinder commented on samples/net/full_mqtt_tls/README.rst:
```diff
@@ -0,0 +1,44 @@
+********************
+This application has been built and tested on the NXP FRDMK64F.  RSA or
+ECDSA certs/keys are required to authenticate to the Google IOT Cloud.
+The application includes a key creation script.  The create_keys.py script
```
Better would be:

    Run the ``create_keys.py`` script in the
    ``samples/net/full_mqtt_tls/src/private_info/`` directory.

agross-linaro replied:
done

dbkinder commented on samples/net/full_mqtt_tls/README.rst:
```diff
@@ -0,0 +1,44 @@
+
+Users will also be required to configure the following Kconfig options
+based on their Google Cloud IOT project:
+- CLOUD_CLIENT_ID - created from <telemetry>/<region>/registries/<registry name>/devices/<device name>
```
Add a blank line before the first item in a list (otherwise it's not recognized as a list, and will be included with the previous sentence.)

agross-linaro replied:
done

agross-linaro commented:
re-working now that some of these other PRs have merged.

Commit:
samples: net: Add Google IOT Cloud full stack application

This application provides a full stack implementation for
communicating with the Google IOT Cloud.  It leverages the
a JWT library to create JSON tokens which are then signed using
a RSA or ECDSA private key.  The signed token is used to
authenticate with the Google IOT cloud servers.  Once that is
accomplished, the application can subscribe, or publish data at
will.

Included in the application is a python script that can be used to
create RSA/ECDSA certificates and keys.  The user will load the
certificate for the appropriate device on the Google IOT Cloud.  The
private key is created by the script and stored in a C structure format
that is then consumed by the application.  It is important to note that
any keys generated from the scripts SHOULD NOT BE STORED publicly.

Please refer to the README for more information.

Signed-off-by: David Brown <david.brown@linaro.org>
Signed-off-by: Andy Gross <andy.gross@linaro.org>

Commit:
ext: crypto: mbedtls: Fix uninitialized variable

This patch fixes an issue with an uninitialized variable in the x509
mbedtls feature.  I sent a related patch to the mbedtls project so
that this can be fixed in the future.

https://github.com/ARMmbed/mbedtls/pull/2392

Signed-off-by: Andy Gross <andy.gross@linaro.org>

agross-linaro commented:
> > re-working now that some of these other PRs have merged.
> 
> I have found this PR and am very interested to lend my support.
> 
> I have a MIMXRT1050 Dev Kit and happy to test your developments against it.

try it out if you get a chance.  if you have any issues with the right configuration, let me know and maybe i can point you in the right direction.

agross-linaro commented:
recheck

pfalcon reviewed (commented):
>> Could you separate the JWT part from this sample, meaning that submitting the JWT in separate PR?

> What benefit is there to doing this? Nothing yet depends on JWT, and github is pretty terrible at dealing with dependent pull requests.

I can't find who made the original suggestion via browser page search, but I would be +1 on that, we could have merged that part long ago. As long as it's generally useful library, we can just offer it to users to use. An example in lvgl support recently merged. Nothing depends on it, it's just there for users to leverage.

Skipping that matter, why jwt library was added as subsys/jwt/ ? You see, even the tests included here are tests/lib/jwt/ . The right location for it would be lib/jwt/ .

Finally, about the header name. Everyone who would write JWT library would call header for it `jwt.h`. And that's exactly why we shouldn't do that. We have issues on those matters pending, as they actually starting to hit: https://github.com/zephyrproject-rtos/zephyr/issues/12129 , https://github.com/zephyrproject-rtos/zephyr/issues/12935 . So, it would be time to think about those matters for new components we add.

agross-linaro commented:
Separating out the jwt patch from this push is *NOT* going to happen.

agross-linaro commented:
@pfalcon I can move it to lib.

What is the significance of #12935 to me on this ticket?

agross-linaro commented:
@pfalcon 
Now I remember why it is in subsys and not lib.  Lib is built before ext.  Ext contains mbedtls, and mbedtls is required as a dependency for jwt.  So we can't reorder this unfortunately unless we move ext higher up.

pfalcon reviewed (approved):
Let's get this into 1.14.

agross-linaro commented:
@pfalcon I fail to understand the arbitrary nature of the include file change.  From reading the first ticket you posted, this would involve moving this to include/zephyr.

pfalcon commented:
> I fail to understand the arbitrary nature of the include file change. From reading the first ticket you posted, this would involve moving this to include/zephyr.

Well, I don't insist. I'm "raising awareness" of the predictable, and now actually happening, issues. If we wait until each of us face a particular issue, and only then start to discuss it, it well, will take years, and cause each of us a lot of frustration first.

Anyway, if the move of all headers under `zephyr/` will happen, so this header will become `zephyr/jwt.h`, that arguably will solve the issue too (so, feel free to +1 that ticket).

agross-linaro commented:
@pfalcon so i moved the test from lib to subsys, renamed the app, and fixed up the include to be in include/zephyr

d3zd3z commented:
As far as mentioning "Google", I would suggest it be mentioned as a Google-specific test in the commits, but there is only a small part of the code that is specific to Google. Ideally, future patches can easily make this work with other cloud providers.

In other words, please keep it under a directory like it is so it is easy to make more general.

pfalcon commented:
> In other words, please keep it under a directory like it is so it is easy to make more general.

Well, @agross-linaro already moved it to google_iot_mqtt/Kconfig, and I think that's good move. That's usual problem of general and generic solution few people care about vs a too-specific solution people find miss their usecase (but other people find it hits right on the spot).

Speaking of "more general", we already have a generic MQTT sample. What we've been missing all this time is a show-off that Zephyr can work with real-world cloud providers. I always thought the purpose of this sample is to be exactly that, and then let's not disguise it as something else. (If someone will want to port it to AWS or something, they can, sure.)

d3zd3z commented:
Ok, but what is stopping us from merging this soon? It isn't very useful sitting out of tree, and generalization can certainly take place later.

agross-linaro commented:
@dbkinder @jukkar addressed concerns in subsequent rework.

pfalcon commented:
> Ok, but what is stopping us from merging this soon?

Only getting rid of -1's in reviews.

jukkar commented on samples/net/google_iot_mqtt/src/globalsign.inc:
```diff
@@ -0,0 +1,107 @@
+/*
```
We have a way to automatically generate the .inc file, see for example `samples/net/sockets/echo_server/CMakeLists.txt` how to do it. That way it is much easier to replace the cert when needed.

agross-linaro replied:
so basically if i have a .der encoded file it'll generate the rest.  interesting

jukkar replied:
Did you investigated how the .inc file could be generated automatically? This is not a blocker for this PR but might be good to have in the future.

jukkar commented on samples/net/google_iot_mqtt/src/dhcp.c:
```diff
@@ -0,0 +1,70 @@
+
+	iface = net_if_get_default();
+
+	net_dhcpv4_start(iface);
```
The iface parameter seems useless as you are only using it once, you could just do `net_dhcpv4_start(net_if_get_default());`

agross-linaro replied:
fixed

jukkar commented on samples/net/google_iot_mqtt/src/main.c:
```diff
@@ -0,0 +1,196 @@
+ * time_t defined by newlib is a signed 32-bit value, and will
+ * overflow in 2037.
+ */
+time_t k_time(time_t *ptr)
```
You are stealing the kernel namespace here. Why not call it my_k_time() or similar instead?

agross-linaro replied:
changed and moved to protocol.c

jukkar commented on samples/net/google_iot_mqtt/src/private_info/.gitignore:
```diff
@@ -0,0 +1,4 @@
+*-cert.pem
```
These are not needed if you put the pem/der files to version control and generate the .inc file automatically from them.

agross-linaro replied:
This would be a security violation as you'd be checking in a private key that is unique to a user.

jukkar commented on samples/net/google_iot_mqtt/src/private_info/create_keys.py:
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
```
Not sure if this is really needed if we generate .inc file automatically. The openssl commands could be described in readme file instead.

agross-linaro replied:
we need this for people to generate their own private keys that they can then load to the google iot cloud, or where ever their broker is.

jukkar reviewed (commented):
Looks good, just minor comments from me.

rlubos commented on samples/net/google_iot_mqtt/src/protocol.c:
```diff
@@ -0,0 +1,370 @@
+			u8_t d[32];
+			int len = evt->param.publish.message.payload.len;
+
+			mqtt_read_publish_payload(&client_ctx, d, 32);
```
There's a slight issue with this part, you should verify how much payload data did you actually recieve. Underlying, it's a non-blocking `recv` call, it might return with no data or not complete data. There's also no guarantee that `sizeof(d)` is larger than payload `len` (though it might be correct in this specific case).

See how it's implemented in one of the MQTT tests:
https://github.com/zephyrproject-rtos/zephyr/blob/d646eff3d37dbcf7cac5821d65f3d1a9dafefcc2/tests/net/lib/mqtt_pubsub/src/test_mqtt_pubsub.c#L118
Note that it's a testcase, hence it has some extra checks. But in general, you should read the payload, until you read `len` bytes of data.

pfalcon replied:
> But in general, you should read the payload, until you read len bytes of data.

Yes, that's an absolute requirement, or protocol stream will get unsynchronized. @rlubos, we don't have checks for that, do we?

> Underlying, it's a non-blocking recv call, it might return with no data or not complete data.

Btw, by the end of review of your original PR I got an idea that we need to provide a way to switch MQTT connection into blocking mode, just for the duration of user's reading out the message. I didn't raise it then, to not delay that PR further, but I'm sure we'll need to do that ;-).

agross-linaro replied:
wait what...... so you're telling me I can receive a publish message and have my event function called, and the mqtt hasn't read out the complete payload?

pfalcon replied:
> have my event function called, and the mqtt hasn't read out the complete payload?

Of course not, how mqtt can read the complete payload, if it's 256MB? It's up to user of the API to deal with all those potential megabytes. Streaming support is the key requirement in Zephyr networking APIs.

agross-linaro replied:
Fun.  It's a great point of failure too.

agross-linaro replied:
fixed

rlubos replied:
> I can receive a publish message and have my event function called, and we might not have received the full MQTT message

@agross That's how it works now, as @pfalcon says. Main concern here was payload size, we do not limit it now by internal MQTT buffers.

> Yes, that's an absolute requirement, or protocol stream will get unsynchronized. @rlubos, we don't have checks for that, do we?

`mqtt_input` will not allow you to readout a "new" message until you read entire payload of the previous message from a socket, it'll return an error code in that case.

> Btw, by the end of review of your original PR I got an idea that we need to provide a way to switch MQTT connection into blocking mode

Seeing this MQTT implementation used in a few applications now I came up the same conclusion. Having to use `poll` is quite cumbersome in this case.

rlubos reviewed (approved):
Briefly looked through the code, MQTT stuff looks fine 👍

agross-linaro commented:
rebased to fix doc merge conflict

pfalcon commented:
@dbkinder : Only docs approval is needed for this long-standing PR, can you please have a look?

agross-linaro commented:
rebased and fixed sntp api change

dbkinder reviewed (approved):
+1 for docs, thanks!
